test/unity/toolkits/object_detection/test_object_detector.cxx

```diff
@@ -511,7 +565,81 @@ BOOST_AUTO_TEST_CASE(test_object_detector_train) {
   // Deconstructing `model` here will assert that every expected call to a
   // mocked-out method has been called.
 }
+
+BOOST_AUTO_TEST_CASE(test_object_detector_auto) { 
+
+  test_object_detector model;
+
+  std::unique_ptr<mock_model_backend> mock_nn_model(
+      new mock_model_backend);
+  std::unique_ptr<mock_compute_context> mock_context(new mock_compute_context);
+
+  static constexpr size_t test_max_iterations = 1;
+  static constexpr size_t test_batch_size = 2;
+  const std::vector<std::string> test_class_labels = { "label1", "label2"};
+  static constexpr size_t test_num_examples = 200;
+
+  const std::string test_mlmodel_path = "/test/foo.mlmodel";
+  const std::string test_annotations_name = "test_annotations";
+  const std::string test_image_name = "test_image";
+
+  size_t num_perform_evalution_calls = 0;
+  size_t perform_evaluation_cumulative_data_size = 0;
+  auto perform_evaluation_impl = [&](gl_sframe data, std::string metric) {
+    perform_evaluation_cumulative_data_size += data.size();
+    if (++num_perform_evalution_calls == 2) {
+      TS_ASSERT_EQUALS(perform_evaluation_cumulative_data_size, test_num_examples);
+    }
+    std::map<std::string, variant_type> result;
+    result["mean_average_precision"] = 0.80f;
+    return result;
+  };
+
+  model.perform_evaluation_calls_.resize(2, perform_evaluation_impl);
+
+
+
+  auto perform_training_iteration_impl = []{};
+  model.perform_training_iteration_calls_.push_back(perform_training_iteration_impl);
+
+  
+
+  auto init_train_impl = [&](gl_sframe data, std::string annotations_column_name,
+                          std::string image_column_name,
+                          std::map<std::string, flexible_type> opts) {
+    std::unique_ptr<model_spec> nn_spec(new model_spec);
+    nn_spec->add_convolution("test_layer", "test_input", 16, 16, 3, 3, 1, 1,
+                             model_spec::padding_type::SAME,
+                             /* weight_init_fn */ [](float*w , float* w_end) {
+                               for (int i = 0; i < w_end - w; ++i) {
+                                 w[i] = static_cast<float>(i);
+                               }
+                             });
+
+    float_array_map trained_weights = model.export_weights();
     
+    nn_spec->update_params(trained_weights);
```

Comment:
It looks like you're trying to address the fact that `object_detector::train` depends on `nn_spec_` being non-`NULL` (as well as `training_model_`), after the call to `init_train`. So we just need to set these data members manually in `init_train_impl`. Unfortunately, to this end we should probably add protected setters to the base class, e.g. `set_nn_spec` and `set_training_model`... for now. (In the longer term, I think the class can be refactored a bit, so that e.g. `init_train` returns all the things that `init_train` actually initializes.)

Then the body of `init_train_impl` looks something more like:
```
model.set_training_model(std::unique_ptr<model_backend>(new mock_model_backend));
model.set_nn_spec(std::unique_ptr<nn_spec>(new nn_spec));
```

Note that all `object_detector::train` does with `training_model_` is call `export_weights`, and a default `mock_model_backend` already returns an empty `float_array_map`, which is fine.

Then `object_detector::train` will call `nn_spec_->update_params` with an empty `float_array_map`, which should also be fine.